### ISSUE TITLE
Cannot get @AutoValue working with AspectJ 

### ISSUE THREAD

**Issue by sedubois** [google/auto#210]
Hi, I have a Maven project which depends on aspectj-maven-plugin 1.7, aspectjrt + aspectjtools 1.8.5, and spring-aspects 4.1.5 (thus using AspectJ with compile-time weaving). I then added a dependency on auto-value 1.1-SNAPSHOT and annotated one value object with @AutoValue, but then `mvn clean install -U` yields this error:

    [INFO] --- aspectj-maven-plugin:1.7:compile (default) @ dreamapp-common ---
    [INFO] Showing AJC message detail for messages of types: [error, warning, fail]
    [ERROR] @AutoValue processor threw an exception: java.lang.NullPointerException
	at org.aspectj.org.eclipse.jdt.internal.compiler.lookup.SourceTypeBinding.sourceStart(SourceTypeBinding.java:2408)
	at org.aspectj.org.eclipse.jdt.internal.compiler.lookup.MethodBinding.sourceStart(MethodBinding.java:1270)
	at org.aspectj.org.eclipse.jdt.internal.compiler.apt.model.TypeElementImpl$SourceLocationComparator.determineSourceStart(TypeElementImpl.java:99)
	at org.aspectj.org.eclipse.jdt.internal.compiler.apt.model.TypeElementImpl$SourceLocationComparator.getSourceStart(TypeElementImpl.java:72)
	at org.aspectj.org.eclipse.jdt.internal.compiler.apt.model.TypeElementImpl$SourceLocationComparator.compare(TypeElementImpl.java:65)
	at org.aspectj.org.eclipse.jdt.internal.compiler.apt.model.TypeElementImpl$SourceLocationComparator.compare(TypeElementImpl.java:1)
	at java.util.TimSort.countRunAndMakeAscending(TimSort.java:351)
	at java.util.TimSort.sort(TimSort.java:216)
	at java.util.Arrays.sort(Arrays.java:1512)
	at java.util.ArrayList.sort(ArrayList.java:1454)
	at java.util.Collections.sort(Collections.java:175)
	at org.aspectj.org.eclipse.jdt.internal.compiler.apt.model.TypeElementImpl.getEnclosedElements(TypeElementImpl.java:166)
	at com.google.auto.value.processor.AutoValueProcessor.findLocalAndInheritedMethods(AutoValueProcessor.java:323)
	at com.google.auto.value.processor.AutoValueProcessor.findLocalAndInheritedMethods(AutoValueProcessor.java:317)
	at com.google.auto.value.processor.AutoValueProcessor.defineVarsForType(AutoValueProcessor.java:382)
	at com.google.auto.value.processor.AutoValueProcessor.processType(AutoValueProcessor.java:369)
	at com.google.auto.value.processor.AutoValueProcessor.process(AutoValueProcessor.java:127)
	at org.aspectj.org.eclipse.jdt.internal.compiler.apt.dispatch.RoundDispatcher.handleProcessor(RoundDispatcher.java:140)
	at org.aspectj.org.eclipse.jdt.internal.compiler.apt.dispatch.RoundDispatcher.round(RoundDispatcher.java:122)
	at org.aspectj.org.eclipse.jdt.internal.compiler.apt.dispatch.BaseAnnotationProcessorManager.processAnnotations(BaseAnnotationProcessorManager.java:160)
	at org.aspectj.org.eclipse.jdt.internal.compiler.Compiler.processAnnotations(Compiler.java:844)
	at org.aspectj.org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:439)
	at org.aspectj.org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:420)
	at org.aspectj.ajdt.internal.core.builder.AjBuildManager.performCompilation(AjBuildManager.java:1036)
	at org.aspectj.ajdt.internal.core.builder.AjBuildManager.performBuild(AjBuildManager.java:272)
	at org.aspectj.ajdt.internal.core.builder.AjBuildManager.batchBuild(AjBuildManager.java:185)
	at org.aspectj.ajdt.ajc.AjdtCommand.doCommand(AjdtCommand.java:112)
	at org.aspectj.ajdt.ajc.AjdtCommand.runCommand(AjdtCommand.java:60)
	at org.aspectj.tools.ajc.Main.run(Main.java:371)
	at org.aspectj.tools.ajc.Main.runMain(Main.java:248)
	at org.codehaus.mojo.aspectj.AbstractAjcCompiler.execute(AbstractAjcCompiler.java:524)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:132)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:120)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:355)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:155)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:584)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:216)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:160)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
	at org.codehaus.classworlds.Launcher.main(Launcher.java:47)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:483)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:134)

As a workaround, I can add the annotated class to aspectj-maven-plugin's `<excludes>` and then it compiles fine again, but this is hardly a maintainable solution.

I'm quite new to AspectJ and annotation processors, so don't know if the issue is with the auto-value library, the aspectj library or with my setup. Do you have any suggestion?

Below is my root POM for information. Thanks a lot!

    <project xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
             xmlns="http://maven.apache.org/POM/4.0.0"
             xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">

      <modelVersion>4.0.0</modelVersion>
      <groupId>com.dreamapp</groupId>
      <artifactId>dreamapp-root</artifactId>
      <version>1.0-SNAPSHOT</version>
      <packaging>pom</packaging>

      <modules>
        <module>common</module>
        ...
      </modules>

      <properties>
        <sourceJavaVersion>1.8</sourceJavaVersion>
        <targetJavaVersion>1.8</targetJavaVersion>
        <springVersion>4.1.5.RELEASE</springVersion>
        <aspectjVersion>1.8.5</aspectjVersion>
        ...
      </properties>

      <build>
        <plugins>
          <plugin>
            <artifactId>maven-compiler-plugin</artifactId>
            <version>3.2</version>
            <configuration>
              <source>${sourceJavaVersion}</source>
              <target>${targetJavaVersion}</target>
              <encoding>UTF-8</encoding>
            </configuration>
          </plugin>
          <plugin>
            <groupId>org.codehaus.mojo</groupId>
            <artifactId>aspectj-maven-plugin</artifactId>
            <version>1.7</version>
            <dependencies>
              <dependency>
                <groupId>org.aspectj</groupId>
                <artifactId>aspectjrt</artifactId>
                <version>${aspectjVersion}</version>
              </dependency>
              <dependency>
                <groupId>org.aspectj</groupId>
                <artifactId>aspectjtools</artifactId>
                <version>${aspectjVersion}</version>
              </dependency>
            </dependencies>
            <configuration>
              <showWeaveInfo>true</showWeaveInfo>
              <source>${sourceJavaVersion}</source>
              <target>${targetJavaVersion}</target>
              <complianceLevel>${sourceJavaVersion}</complianceLevel>
              <encoding>UTF-8</encoding>
              <aspectLibraries>
                <aspectLibrary>
                  <groupId>org.springframework</groupId>
                  <artifactId>spring-aspects</artifactId>
                </aspectLibrary>
              </aspectLibraries>
              
              <Xlint>ignore</Xlint>
            </configuration>
            <executions>
              <execution>
                <goals>
                  <goal>compile</goal>
                  <goal>test-compile</goal>
                </goals>
              </execution>
            </executions>
          </plugin>
        </plugins>
      </build>

      <repositories>
        <repository>
          <id>sonatype</id>
          <name>sonatype</name>
          <url>https://oss.sonatype.org/content/groups/public</url>
          <snapshots>
            <enabled>true</enabled>
          </snapshots>
          <releases>
            <enabled>true</enabled>
          </releases>
        </repository>
      </repositories>

      <dependencies>
        ...
        <dependency>
          <groupId>com.google.auto.value</groupId>
          <artifactId>auto-value</artifactId>
          <version>1.1-SNAPSHOT</version>
          <scope>provided</scope>
        </dependency>
        <dependency>
          <groupId>org.aspectj</groupId>
          <artifactId>aspectjrt</artifactId>
          <version>${aspectjVersion}</version>
        </dependency>
        <dependency>
          <groupId>org.aspectj</groupId>
          <artifactId>aspectjtools</artifactId>
          <version>${aspectjVersion}</version>
        </dependency>
        <dependency>
          <groupId>org.springframework</groupId>
          <artifactId>spring-aspects</artifactId>
          <version>${springVersion}</version>
        </dependency>
        ...
      </dependencies>
    </project>

**Comment by tbroyer** [comment]
```
at org.aspectj.org.eclipse.jdt.internal.compiler.apt.model.TypeElementImpl.getEnclosedElements(TypeElementImpl.java:166)
at com.google.auto.value.processor.AutoValueProcessor.findLocalAndInheritedMethods(AutoValueProcessor.java:323)
at com.google.auto.value.processor.AutoValueProcessor.findLocalAndInheritedMethods(AutoValueProcessor.java:317)
at com.google.auto.value.processor.AutoValueProcessor.defineVarsForType(AutoValueProcessor.java:382)
at com.google.auto.value.processor.AutoValueProcessor.processType(AutoValueProcessor.java:369)
at com.google.auto.value.processor.AutoValueProcessor.process(AutoValueProcessor.java:127)
at org.aspectj.org.eclipse.jdt.internal.compiler.apt.dispatch.RoundDispatcher.handleProcessor(RoundDispatcher.java:140)
```

Seeing this in the stack trace, I'd say it's a bug in AspectJ.

As a workaround you could possibly run a first pass of maven-compiler-plugin with `<proc>only</proc>`, or processor-maven-plugin; then the aspectj-maven-plugin with `<proc>none</proc>`

**Comment by sedubois** [comment]
Thanks for the idea, but then I think I prefer the workaround I mentioned, at least I would need to change my POM only every time I write @AutoValue in a new place.

As you suggested, I opened an AspectJ bug: https://bugs.eclipse.org/bugs/show_bug.cgi?id=462782

**Comment by eamonnmcmanus** [comment]
I think it's suspicious that the exception is in org.aspectj.org.eclipse.jdt.internal.compiler.lookup.SourceTypeBinding. Does the initial org.aspectj mean that AspectJ has run on the internal Eclipse classes? You probably don't want that.

**Comment by eamonnmcmanus** [comment]
Hmm, I think the org.aspectj prefix is probably just shading, since it's there all the way through. But on searching for "AspectJ annotation processing" I find that support for annotation processors is [quite recent](https://eclipse.org/aspectj/doc/released/README-182.html) and they say (as of last August) "There is still work to be done to get annotation processors behaving under AJDT." So it might indeed be that you will need to keep AspectJ away from any class that annotation processors are likely to run on.

**Comment by sedubois** [comment]
OK so looks like this is unrelated to auto-value. Let's hope the bug I raised in Eclipse helps to get this fixed. Thanks for the feedback and keep up the good work, I'm impatient to see a stable 1.1 release to get @AutoValue.Builder :)

**Comment by sedubois** [comment]
There has been progress on the AspectJ side, cfr. https://bugs.eclipse.org/bugs/show_bug.cgi?id=462782#c4. Is the reported error an AutoValue issue?

**Comment by eamonnmcmanus** [comment]
I replied in https://bugs.eclipse.org/bugs/show_bug.cgi?id=462782#c6. Basically the issue now is a conflict between versions of Apache Velocity.

**Comment by sedubois** [comment]
Should the AspectJ bug be closed and the velocity shading issue be further tracked here?

**Comment by eamonnmcmanus** [comment]
This should be fixed by #259, which replaces Apache Velocity with a subset reimplementation.
